features/ForceExpiration.php

```diff
@@ -117,43 +157,36 @@ private function _handleNotification(User $user) : void {
 				$notificationDao->delete($notification);
 			}
 		}
-
-		$notification = $notificationManager->createTrivialNotification($user->getId(), NOTIFICATION_TYPE_WARNING, array('contents' => __('plugins.generic.betterPassword.message.yourPasswordWillExpire', ['days' => $diffInDays])));
+		$notification = $notificationManager->createTrivialNotification($user->getId(), \PKP\notification\PKPNotification::NOTIFICATION_TYPE_WARNING, array('contents' => __('plugins.generic.betterPassword.message.yourPasswordWillExpire', ['days' => $diffInDays])));
 		$lastNotification = (object) ['id' => $notification->getId(), 'days' => $diffInDays];
-		
 		$this->_setLastNotification($user, $lastNotification);
-		/** @var UserDAO */
-		//$userDao = DAORegistry::getDAO('UserDAO');
-		//$userDao->updateObject($user);
-		$userDao = Repo::user()->dao;
-		$userDao->edit($user);
-		
+		$user = Repo::user()->edit($user);
 	}
 
 	/**
 	 * Retrieve if the password is expired
 	 * @param User $user
 	 * @return bool
 	 */
-	private function _isPasswordExpired(User $user) : bool {
+	private function _isPasswordExpired(\PKP\user\User $user) : bool {
 		return $this->_getExpirationDate($user)->getTimestamp() <= time();
 	}
 
 	/**
 	 * Retrieve the last notification
 	 * @param User $user
-	 * @return ?string
+	 * @return ?object Last Notification
 	 */
-	private function _getLastNotification(User $user) : ?object {
+	private function _getLastNotification(\PKP\user\User $user) : ?object {
```

Comment:
How is this function working (or is it)?  It appears to be using User settings, which I thought we were unable to save.

Reply:
It is currently working, and the data is being saved in the notifications table (possibly due to the notification manager). However lastPasswordNotification is not being saved to the user database.

Reply:
Saving the notification itself is done by line 160:
```
$notification = $notificationManager->createTrivialNotification($user->getId(), \PKP\notification\PKPNotification::NOTIFICATION_TYPE_WARNING, array('contents' => __('plugins.generic.betterPassword.message.yourPasswordWillExpire', ['days' => $diffInDays])));
```

Saving (and then fetching) the setting within the User object is unable to occur with:
```
$user->setData("{$this->_plugin->getSettingsName()}::lastPasswordNotification", json_encode($notification));
```
```
return json_decode($user->getData("{$this->_plugin->getSettingsName()}::lastPasswordNotification"), false);
```
because of our inability to hook the User schema.